libcxx/test/std/input.output/file.streams/fstreams/filebuf.members/close.pass.cpp

```diff
@@ -37,6 +41,13 @@ int main(int, char**)
         assert(f.close() == nullptr);
         assert(!f.is_open());
     }
+#if defined(__BIONIC__)
+    // Starting with Android API 30+, Bionic's fdsan aborts a process that
+    // attempts to close a file descriptor belonging to something else. Disable
+    // fdsan to allow closing the FD belonging to std::filebuf's FILE*.
+    if (android_fdsan_set_error_level != nullptr)
+        android_fdsan_set_error_level(ANDROID_FDSAN_ERROR_LEVEL_DISABLED);
+#endif
 #if defined(__unix__)
     {
         std::filebuf f;
```

Comment:
Is what we're doing in this test Standards conforming? If so, is there a reason why fdsan flags it? That would seem like a false positive to me.

Reply:
`std::filebuf::__open` isn't in the C++ standard, so that would make the the test non-conforming for C++.

Maybe there's a POSIX conformance question for the combination of libc++ and this test? Sometimes programs close the default input/output FDs (e.g. `STDOUT_FILENO`), even though those FDs are associated with `FILE*` (e.g. `stdout`). fdsan has special handling to accommodate the standard streams, https://android.googlesource.com/platform/bionic/+/refs/tags/android-14.0.0_r61/libc/stdio/stdio.cpp#110.

`__open` is using `fdopen`, and Bionic's `fdopen` uses `android_fdsan_exchange_owner_tag` here, https://android.googlesource.com/platform/bionic/+/refs/tags/android-14.0.0_r61/libc/stdio/stdio.cpp#237:

```
  android_fdsan_exchange_owner_tag(fd, 0, __get_file_tag(fp));
```

This line asserts that the FD's existing tag is 0 (i.e. that the FD is unowned), and then marks the FD as owned by the new `FILE*`. Hence, the `close()` call in the libc++ test fails, because it doesn't provide the expected FD tag. I believe Bionic's behavior is intentional here.

This test is intentionally doing a double-close, which is the exact thing that fdsan was created to detect. https://android.googlesource.com/platform/bionic/+/master/docs/fdsan.md#background

For this test, I'd also be OK with adding `!defined(__BIONIC__)` to `defined(__unix__)`, with a short comment about why Bionic is special.

Reply:
Yes, I'd much rather simply skip this test on Bionic. Re-reading the test, it is indeed doing something wrong but it's guarding with `#if defined(__unix__)`, which basically means "I know I'm doing something wrong, but defined(__unix__) lets me get away with it". Instead, we want to change that to "I know I'm doing something wrong, but defined(__unix__) && !defined(__BIONIC__) lets me get away with it", since Bionic doesn't let you get away with it.